### PR TITLE
added notebook with FairPrep paper results

### PR DESCRIPTION
**Pull request by vvthakral**

### CODE REVIEW

**Comment by sscdotopen** [comment]
Hi Vishnu,

This looks good, but it needs a bit more polishing. Could you add a PR with the example in a .py file? Unfortunately, its very difficult to comment on the notebook.

File: examples/fair_prep_results.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+import os
+os.chdir("..")
```
Comment (sscdotopen) [inline examples/fair_prep_results.py:5]: This should not be here. Make sure that the file can be run from the root dir of the repository

File: examples/fair_prep_results.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+import shutil
+import warnings
+warnings.simplefilter('ignore')
+
+from fp.traindata_samplers import CompleteData
+from fp.missingvalue_handlers import CompleteCaseAnalysis
+from fp.dataset_experiments import GermanCreditDatasetSexExperiment
+from fp.dataset_experiments import RicciRaceExperiment
```
Comment (sscdotopen) [inline examples/fair_prep_results.py:13]: Why is this imported?

File: examples/fair_prep_results.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+samplers = [CompleteData()]
+missings = [CompleteCaseAnalysis()]
+scalers = [NamedMinMaxScaler(), NamedStandardScaler()]
+learners = [NonTunedLogisticRegression(), LogisticRegression(), NonTunedDecisionTree(), DecisionTree()]
+post_processors = [NoPostProcessing(),RejectOptionPostProcessing(), CalibratedEqualOddsPostProcessing()]
+pre_processors = [NoPreProcessing(), DIRemover(1.0), DIRemover(0.5), Reweighing()]
+
+path = os.path.join(os.getcwd(), "logs")
+
+def extract_info(present_acc,present_di,present_fnr,present_fpr):
```
Comment (sscdotopen) [inline examples/fair_prep_results.py:33]: please refactor this into a helper file, it should not be part of the notebook

File: examples/fair_prep_results.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    '''
+    Function to extract the metrics (accuracy, DI, false_positive_rate, false_negative_rate)value from the log files 
+    '''
+    try:
+        for folderName, subfolders, filenames in os.walk(path):
```
Comment (sscdotopen) [inline examples/fair_prep_results.py:38]: please don't use camel case naming for variables in python, follow idiomatic naming, e.g. folder_name here

File: examples/fair_prep_results.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+def extract_info(present_acc,present_di,present_fnr,present_fpr):
+    '''
+    Function to extract the metrics (accuracy, DI, false_positive_rate, false_negative_rate)value from the log files 
+    '''
+    try:
+        for folderName, subfolders, filenames in os.walk(path):      
+            for subfolder in subfolders:
+                filelist = os.listdir(path+'/'+subfolder)
+                for f in filelist:
```
Comment (sscdotopen) [inline examples/fair_prep_results.py:41]: give readable names to variable like file not f

File: examples/fair_prep_results.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+                    file_path = path+'/'+subfolder+'/'+f
+                    with open(file_path,'r+') as f:
```
Comment (sscdotopen) [inline examples/fair_prep_results.py:43]: please read the file as a pandas dataframe and use pandas to retrieve the values, that makes the code shorter, easier to read and more robust

File: examples/fair_prep_results.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+                            
+                            elif line.startswith("test") and 'generalized_true_positive_rate' in line and 'None' in line:
+                                fpr = line.split(',')
+                                fpr_val = fpr[-1].rstrip("\n")
+                                present_fpr.append(float(fpr_val))
+                shutil.rmtree(path+'/'+subfolder)            
+    except:
+        pass
+    
+    return present_acc,present_di,present_fnr,present_fpr
```
Comment (sscdotopen) [inline examples/fair_prep_results.py:69]: spaces between variables

File: examples/fair_prep_results.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+def exp_run_pre_pro_param(seed,learn,pre,post):
```
Comment (sscdotopen) [inline examples/fair_prep_results.py:71]: give understandable names to functions please

File: examples/fair_prep_results.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+    '''
+    Experiment function to run the experiments
+    '''
+    exp = GermanCreditDatasetSexExperiment(
```
Comment (sscdotopen) [inline examples/fair_prep_results.py:75]: indent the parameters

File: examples/fair_prep_results.py [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+    numeric_attribute_scaler=NamedStandardScaler(),
+    learners=[learn],
+    pre_processors=[pre],
+    post_processors=[post]
+    )
+    
+    exp.run()
+
+#Mergining processors to one format for input
+processors = [[NoPreProcessing(),NoPostProcessing()], [DIRemover(1.0),NoPostProcessing()], [DIRemover(0.5),NoPostProcessing()], [Reweighing(),NoPostProcessing()],
```
Comment (sscdotopen) [inline examples/fair_prep_results.py:88]: please use an array of tuples, not an array of arrays

File: examples/fair_prep_results.py [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+from fp.scalers import NamedStandardScaler, NamedMinMaxScaler
+from fp.learners import NonTunedLogisticRegression, LogisticRegression, DecisionTree, NonTunedLogisticRegression, NonTunedDecisionTree          
+from fp.post_processors import NoPostProcessing,RejectOptionPostProcessing, EqualOddsPostProcessing, CalibratedEqualOddsPostProcessing
+from fp.pre_processors import NoPreProcessing, DIRemover, Reweighing
+
+import numpy as np
+get_ipython().run_line_magic('matplotlib', 'inline')
+import matplotlib.pyplot as plt
+
+seeds = [0xbeef, 0xcafe, 0xdead, 0xdeadcafe, 0xdeadbeef, 0xbeefcafe, 0xcafebeef, 50, 583, 5278, 100000, 0xefac,0xfeeb, 0xdaed, 0xefacdaed, 0xfeebdead]
```
Comment (sscdotopen) [inline examples/fair_prep_results.py:23]: why do we need these?

File: examples/fair_prep_results.py [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+
+#Mergining processors to one format for input
+processors = [[NoPreProcessing(),NoPostProcessing()], [DIRemover(1.0),NoPostProcessing()], [DIRemover(0.5),NoPostProcessing()], [Reweighing(),NoPostProcessing()],
+              [NoPreProcessing(), RejectOptionPostProcessing()],[NoPreProcessing(),CalibratedEqualOddsPostProcessing()]]
+
+
+
+accuracy, disp_imp, fpr, fnr  = [], [], [], []
+
+def run_exp(accuracy,disp_imp,fnr,fpr,seeds,learners,processors):
```
Comment (sscdotopen) [inline examples/fair_prep_results.py:95]: spaces

File: examples/fair_prep_results.py [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+    
+    exp.run()
+
+#Mergining processors to one format for input
+processors = [[NoPreProcessing(),NoPostProcessing()], [DIRemover(1.0),NoPostProcessing()], [DIRemover(0.5),NoPostProcessing()], [Reweighing(),NoPostProcessing()],
+              [NoPreProcessing(), RejectOptionPostProcessing()],[NoPreProcessing(),CalibratedEqualOddsPostProcessing()]]
+
+
+
+accuracy, disp_imp, fpr, fnr  = [], [], [], []
```
Comment (sscdotopen) [inline examples/fair_prep_results.py:93]: please initialize each variable on its own line and give them readable names

File: examples/fair_prep_results.py [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+def plotter(p_x, p_y, x_ticks, xlabel, main_title):
+    '''
+    Function to plot various comparison plots.
+    '''
+    title_list = ['NoPreProcessing','DIRemover(1.0)','DIRemover(0.5)','Reweighing','Reject Option','Caliberated Equal Odds']
+    label_list = [['NonTunedLogistic','TunedLogistic'],['NonTunedDecisionTree','TunedDecisionTree']] 
+    fig, axs = plt.subplots(6,2,figsize=((10,20)))
+    axs = axs.flatten()
+    for i in range(0,len(accuracy),2):
+        axs[i//2].scatter(p_x[i], accuracy[i],c='b',marker='o')
```
Comment (sscdotopen) [inline examples/fair_prep_results.py:123]: the computation of i//2 is repeated many times, just compute it once

File: examples/fair_prep_results.py (hunk 2 above)
Comment (vvthakral) [reply examples/fair_prep_results.py:13]: It was done for working with RicciDataset in this file, will do it separately, has been removed for here.

File: examples/fair_prep_results.py (hunk 3 above)
Comment (vvthakral) [reply examples/fair_prep_results.py:33]: Done.

File: examples/fair_prep_results.py (hunk 4 above)
Comment (vvthakral) [reply examples/fair_prep_results.py:38]: Okay, updated.

File: examples/fair_prep_results.py (hunk 5 above)
Comment (vvthakral) [reply examples/fair_prep_results.py:41]: okay

File: examples/fair_prep_results.py (hunk 6 above)
Comment (vvthakral) [reply examples/fair_prep_results.py:43]: Makes sense. Have changed it. Surely much easier to interpret now.

File: examples/fair_prep_results.py (hunk 7 above)
Comment (vvthakral) [reply examples/fair_prep_results.py:69]: done.

File: examples/fair_prep_results.py (hunk 14 above)
Comment (vvthakral) [reply examples/fair_prep_results.py:123]: changed

File: examples/fair_prep_results.py (hunk 12 above)
Comment (vvthakral) [reply examples/fair_prep_results.py:95]: updated

File: examples/fair_prep_results.py (hunk 10 above)
Comment (vvthakral) [reply examples/fair_prep_results.py:88]: changed

File: examples/fair_prep_results.py (hunk 13 above)
Comment (vvthakral) [reply examples/fair_prep_results.py:93]: updated.

File: examples/fair_prep_results.py (hunk 9 above)
Comment (vvthakral) [reply examples/fair_prep_results.py:75]: updated

File: examples/fair_prep_results.py (hunk 8 above)
Comment (vvthakral) [reply examples/fair_prep_results.py:71]: changed.

File: examples/fair_prep_results.py (hunk 11 above)
Comment (vvthakral) [reply examples/fair_prep_results.py:23]: Actually, they are not needed for this run, as we are altering only the learners and processors.
Removed them.

**Pull request closed without merging**